zio-flow/shared/src/main/scala/zio/flow/internal/DurableIndexedStore.scala

```diff
@@ -0,0 +1,120 @@
+    for {
+      //TODO : Only catch "Column Family already exists"
+      colFamHandle <-
+        transactionDB
+          .createColumnFamily(new ColumnFamilyDescriptor(ProtobufCodec.encode(Schema[String])(topic).toArray))
+          .catchSome { case _: RocksDBException =>
+            getColFamilyHandle(topic)
+          }
+          .refineToOrDie[IOException]
+      _ <- transactionDB
+             .put(
+               colFamHandle,
+               ProtobufCodec.encode(Schema[String])("POSITION").toArray,
+               ProtobufCodec.encode(Schema[Long])(0L).toArray
+             )
+             .refineToOrDie[IOException]
+      _ <- colHandles.update(list => colFamHandle :: list)
+    } yield colFamHandle
+
+  override def position(topic: String): IO[IOException, Long] = (for {
+    cfHandle <- getColFamilyHandle(topic)
+    positionBytes <-
+      transactionDB.get(cfHandle, ProtobufCodec.encode(Schema[String])("POSITION").toArray).orDie
+    position <- ZIO.fromEither(ProtobufCodec.decode(Schema[Long])(Chunk.fromArray(positionBytes.get)))
+  } yield position).mapError(s => new IOException(s))
+
+  def getNamespaces(): IO[IOException, Map[Chunk[Byte], ColumnFamilyHandle]] =
+    for {
+      list <- colHandles.get
+      map  <- ZIO.succeed(list.map(cHandle => Chunk.fromArray(cHandle.getName) -> cHandle).toMap)
+    } yield map
+
+  def getColFamilyHandle(topic: String): UIO[ColumnFamilyHandle] =
+    for {
+      namespaces <- getNamespaces().orDie
+      cf         <- ZIO.succeed(namespaces.get(ProtobufCodec.encode(Schema[String])(topic)))
+      handle <- cf match {
+                  case Some(h) => ZIO.succeed(h)
+                  case None    => addTopic(topic).orDie
+                }
+    } yield handle
+
+  override def put(topic: String, value: Chunk[Byte]): IO[IOException, Long] = for {
+    colFam <- getColFamilyHandle(topic)
+    _ <- transactionDB.atomically {
+           Transaction.getForUpdate(
+             colFam,
+             ProtobufCodec.encode(Schema[String])("POSITION").toArray,
+             exclusive = true
+           ) >>= { posBytes =>
+             Transaction
+               .put(
+                 colFam,
+                 ProtobufCodec.encode(Schema[String])("POSITION").toArray,
+                 incPosition(posBytes)
+               ) >>= { _ =>
+               Transaction
+                 .put(
+                   colFam,
+                   incPosition(posBytes),
+                   value.toArray
+                 )
+                 .refineToOrDie[IOException]
+             }
+           }
+         }.refineToOrDie[IOException]
+    newPos <- position(topic)
+  } yield newPos
+
+  private def incPosition(posBytes: Option[Array[Byte]]): Array[Byte] =
+    ProtobufCodec
+      .encode(Schema[Long])(
+        ProtobufCodec.decode(Schema[Long])(Chunk.fromArray(posBytes.get)) match {
+          case Left(error) => throw new IOException(error)
```

Comment:
As this is something that should never happen I would include more info in the exception about it, like the topic name and possible the whole chunk.

Reply:
That's right. Done.